docs/rfd.md

```diff
@@ -0,0 +1,478 @@
+  Failed = 5;
+}
+
+message JobStatus {
+  // The job's logical state.
+  State state = 1;
+  // A human-readable message describing the job's state, or an error message
+  // if the job failed.
+  string message = 2;
+  // The PID of the job's process.
+  int32 pid = 3;
+  // The exit code of the job's process.
+  // Only present if the job is in the Completed state.
+  optional int32 exitCode = 4;
+  // A bitmask equivalent to syscall.WaitStatus, also described in wait(2).
+  // Only present if the job is in the Completed or Failed state.
+  optional uint32 waitStatus = 5;
+  // The time at which the job was started.
+  google.protobuf.Timestamp startTime = 6;
+  // The time at which the job completed (or failed). If the job is still
+  // running, this field will not be present.
+  google.protobuf.Timestamp endTime = 7;
+  // The job's original spec.
+  JobSpec spec = 8;
+}
+
+message CommandSpec {
+  // The command name to run. Required.
+  string command = 1;
+  // The command's arguments, not containing the command name itself.
+  repeated string args = 2;
+  // Optional additional environment variables to set for the command.
+  // These will be merged with the job server's environment variables.
+  repeated string env = 3;
+  // The working directory for the command. If not specified, the job server's
+  // working directory will be used.
+  optional string cwd = 4;
+  // Optional user id to run the command as.
+  // If not specified, the job server's user id will be used.
+  optional int32 uid = 5;
+  // Optional group id to run the command as.
+  // If not specified, the job server's group id will be used.
+  optional int32 gid = 6;
+  // Optional data to be written to the process's stdin pipe when it starts.
+  // If not specified, the process's stdin will be set to /dev/null.
+  optional bytes stdin = 7;
+}
+
+message ProcessOutput {
+  // A chunk of output from the process's combined stdout and stderr streams.
+  //
+  // This is a raw byte stream, and may contain arbitrary binary data,
+  // depending on the command being run. It is up to the client to interpret
+  // the output correctly based on the command being run. No guarantees are
+  // made about the size of each chunk, or the frequency at which they
+  // are sent.
+  bytes output = 1;
+}
+
+message ResourceLimits {
+  // Process CPU limit defined in milli-cores (1000 = 1 core/vcpu)
+  optional int64 cpu = 1;
+  // Process memory limit defined in bytes
+  optional MemoryLimits memory = 2;
+  // Process IO limits for devices
+  optional IODeviceLimits io = 3;
+}
+
+message MemoryLimits {
+  // Memory soft limit defined in bytes. Processes exceeding the limit are
+  // not OOM-killed, but may be throttled by the kernel.
+  optional int64 softLimit = 1;
+  // Memory limit defined in bytes. If exceeded, the process will be OOM-killed.
+  optional int64 limit = 2;
+}
+
+message IODeviceLimits {
+  // Limits for individual devices
+  repeated DeviceLimit devices = 1;
```

Comment:
Feel free to have a single I/O limit on a hardcoded/configurable device ID here (it would be strange for the client would have that information anyway).

Reply:
True - what if the client provides a path like /dev/xyz and the server looks up the ID itself?

Reply:
I'd just go for a static device ID (for this challenge, at least).

Reply:
Would that work, since the IDs are hardware specific?